### PR TITLE
Karis/Amber-Lynn - Edges - Ride Share

### PR DESCRIPTION
**Pull request by kimj42**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer | Instructor Feedback
:------------- | :------------- | ---
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | We rewrote tests to that felt incomplete or were not helpful.
Describe any examples of composition that you encountered in this project, if any | When working on creating net expenditures method; issues arose with super and inheritance of trip data in the testing position. | This isn't quite what we mean by "composition". By composition we mean a _has-a_ or _has-many_ relationship. Examples in this project include "a TripDispatcher _has-many_ Drivers, Passengers and Trips", "a Driver _has-many_ Trips", "a Trip _has-a_ Driver".
Describe the relationship between `User` and `Driver`|  Users can be drivers as well as drivers can also be users. A user can end up having multiple drivers while also drivers can have multiple passengers(users). 
Describe a _nominal test_ that you wrote for this assignment. | The total_revenue method required that we write a nominal test
Describe an _edge case test_ that you wrote for this assignment | If the average rating for a driver was divided by zero - an edge case was created to test this. 
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | Raising errors - how too much or not enough clarity in "before do" portion can screw with your code.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | Don't freak out. Visual aids help see and understand how classes and methods are connected

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        subtotal = (trip.cost.to_f - fee)
+        total += subtotal
+      end
+      total = total * 0.80
+      return total
+    end
+
+    def net_expenditures
+      return (super - total_revenue).round(2)
+    end
```
Comment (droberts-sea) [inline lib/driver.rb:70]: Good use of `super`!

File: lib/user.rb [hunk 2]
```diff
@@ -16,5 +19,25 @@ def initialize(input)
     def add_trip(trip)
       @trips << trip
     end
+
+    def net_expenditures
+      sum = 0
+
+      @trips.each do |trip|
```
Comment (droberts-sea) [inline lib/user.rb:26]: Both these methods will hit an error if you try to run them on a user with an incomplete trip. There are a couple of workarounds for this.

You could explicitly ignore trips with a `nil` cost:

```ruby
trips.each do |trip|
  if trip.cost.nil?
    next
  end
  # ... add to the total ...
end
```

Or even better, you could write a helper method that returns a list of complete trips:

```ruby
def completed_trips
  return @trips.reject { |t| t.end_time.nil? }
end

def net_expenditures
  completed_trips.each do |trip|
    # ... same logic as before ...
  end
end
```

File: specs/user_spec.rb [hunk 3]
```diff
@@ -22,40 +22,80 @@
       expect(@user.trips.length).must_equal 0
     end
 
+
+    it "returns sum of net expenditures of user" do
+
+      trip = RideShare::Trip.new(id: 8, driver: nil, passenger: @user,
+        start_time: Time.parse("2016-08-08 06:20:00 -0700"),
```
Comment (droberts-sea) [inline specs/user_spec.rb:29]: I think you're missing some edge cases that are common to both these `User` methods:

- What happens if the user has no trips?
- What happens if the user has an incomplete trip?

File: lib/driver.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
+require 'pry'
+
+module RideShare
+  class Driver < User
+    class InvalidVinError < ArgumentError
+    end
+
+    class InvalidStatusError < ArgumentError
```
Comment (droberts-sea) [inline lib/driver.rb:8]: I like your use of custom errors here.

File: lib/driver.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        raise ArgumentError, "No trip provided"
+      end
+
+      @driven_trips << trip
+    end
+
+    def average_rating
+      sum = 0.00
+      @driven_trips.each do |trip|
+        sum += trip.rating.to_f
```
Comment (droberts-sea) [inline lib/driver.rb:44]: As with the new methods for `User`, all these `Driver` methods that operate on the collection of driven trips will fail for a driver with an incomplete trip.

File: lib/driver.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    end
+
+    def total_revenue
+      total = 0.00
+      fee = 1.65
+
+      @driven_trips.each do |trip|
+        subtotal = (trip.cost.to_f - fee)
+        total += subtotal
+      end
```
Comment (droberts-sea) [inline lib/driver.rb:63]: You could use enumerables here to make your intention more clear:

```ruby
@driven_trips.map { |t| t.cost.to_f - fee }.sum
```

File: specs/driver_spec.rb [hunk 7]
```diff
@@ -1,104 +1,182 @@
...
+                @driver.add_driven_trip(trip2)
+              end
+
+              it "returns a float" do
+                expect(@driver.total_revenue).must_be_kind_of Float
+              end
+
+              it "returns a zero if no trips provided" do
+                driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV",
+                  vin: "1C9EVBRM0YBC564DZ", status: :AVAILABLE)
```
Comment (droberts-sea) [inline specs/driver_spec.rb:125]: Good job catching this edge case!

File: specs/driver_spec.rb [hunk 8]
```diff
@@ -1,104 +1,182 @@
...
-  end
+                end
 
-  describe "net_expenditures" do
-    # You add tests for the net_expenditures method
-  end
-end
+                it "calculates net expenditure for driver" do
+                  expect(@driver.net_expenditures).must_be_close_to -58.68
+                end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:180]: What if the driver has no trips, or has an incomplete trip?

File: lib/trip_dispatcher.rb [hunk 9]
```diff
@@ -72,5 +100,35 @@ def inspect
     def check_id(id)
       raise ArgumentError, "ID cannot be blank or less than zero. (got #{id})" if id.nil? || id <= 0
     end
+
+    def request_trip(user_id)
+      new_user = find_passenger(user_id)
+      driver = @driver.select {|driver| driver.status == :AVAILABLE}
+
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:107]: I had trouble calling this method. You've got it in the `private` section of this class (see line 98 above), which means it's not accessible from outside.

Once I removed `private`, I found the logic was pretty close, but I ran into a couple of problems:

- On line 106, `@driver` should be `@drivers`
- `.select` returns an array, but you store it in a variable called `driver` (implying just one) and use it like it's an instance of `Driver`. `.find` might be what you're looking for here.
- The `Trip` class doesn't yet know how to handle a `nil` end time

The rest of the logic in this method looks great. My guess is that if you had been able to run this method you would have cruise through the rest of these problems. This would have been a good time to come to an instructor for help.

File: lib/trip.rb [hunk 10]
```diff
@@ -1,14 +1,22 @@
 require 'csv'
+require 'driver'
+require 'pry'
```
Comment (droberts-sea) [inline lib/trip.rb:4]: The `require 'driver'` here prevented me from running this code. It needs to be `require_relative`.

File: lib/trip_dispatcher.rb [hunk 11]
```diff
@@ -72,5 +100,35 @@ def inspect
...
+          end_time: new_end_time,
+          cost: new_cost,
+          rating: new_rating
+        }
+
+      trip = Trip.new(trip_data)
+      driver.add_driven_trip(trip)
+
+      driver.status = :UNAVAILABLE
+      new_user.add_trip(trip)
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:127]: Ideally you would create a new method in driver that both adds this trip to the list and marks the driver as unavailable. Maybe something like `Driver#start_trip`?

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | yes - see inline
**Wave 1** |
Appropriate use of Ruby's `Date`| yes
`Trip` has a helper method to calculate duration| yes
`User` (passenger) has a method to calculate total cost of all trips| yes
Tests for wave 1 | missing some edge cases - see inline
**Wave 2** |
`Driver` inherits from `User` | yes
`Driver` has `add_driven_trip` method | yes
`Driver` has method to calculate average rating | yes
`Driver` has method to calculate net expenditures and it uses `super` | yes
`Driver` has a method to calculate total revenue | yes
Tests for wave 2 | missing some edge cases - see inline
**Wave 3** |
`TripDispatcher` has a new method to create trips | yes
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | see inline
Complex logic was correctly implemented | mostly
Tests for `request_trip` | no
Methods from wave 1 and 2 handle incomplete trips | no
Tests for wave 1 and 2 methods with incomplete trips | no
**Overall** | 

This is a good start. The code you've built for waves 1 and 2 is well-organized, and while you're missing some edge cases from your testing, that's basically what we would expect at this point. A good rule of thumb when testing around arrays, either as input or output, is to test for an array with several elements and one with none.

I am a little disappointed that you didn't finish wave 3, but it looks like you made a strong start on it so I'm not too worried.

Please review the comments I've left inline, and remember to ask for help if you're stuck on something (especially a situation like "we can't run our code"). Keep up the hard work!
